File: Erlang/hoshimeguri1.erl
% $ erlc hoshimeguri1.erl
% $ erl -run hoshimeguri1 main --noshell -data ../data.txt 

-module(hoshimeguri1).
-export([main/0]).

star($A, $W) -> $I;
star($A, $R) -> $H;
star($I, $W) -> $G;
star($I, $R) -> $F;
star($G, $W) -> $E;
star($G, $R) -> $D;
star($E, $W) -> $C;
star($E, $R) -> $B;
star($C, $W) -> $A;
star($C, $R) -> $J;
star($H, $W) -> $C;
star($H, $R) -> $J;
star($J, $W) -> $E;
star($J, $R) -> $B;
star($B, $W) -> $G;
star($B, $R) -> $D;
star($D, $W) -> $I;
star($D, $R) -> $F;
star($F, $W) -> $A;
star($F, $R) -> $H.

solve(Input) ->
  lists:reverse(lists:foldl(fun(C, A) -> [star(hd(A), C)|A] end, [hd(Input)], tl(Input))).

test(Input, Expected) ->
  Actual = solve(Input),
  if
    Actual == Expected -> io:put_chars(".");
    true               -> io:format("""

input:    ~s
expected: ~s
actual:   ~s
""", [Input, Expected, Actual])
  end.

test(Line) ->
  [Input, Expected] = string:tokens(Line, [$ , $\n]),
  test(Input, Expected).

read_lines(Handle) ->
  case file:read_line(Handle) of
    {ok, Line} -> [Line|read_lines(Handle)];
    _Otherwise -> []
  end.

main() ->
  {ok, [[Filename]]} = init:get_argument(data),
  {ok, Handle} = file:open(Filename, read),
  Lines = read_lines(Handle),
  file:close(Handle),
  lists:map(fun test/1, Lines),
  io:nl(),
  halt().

File: Go/hoshimeguri1.go
// ビルド:
//    $ go build hoshimeguri1.go
//
// 実行:
//    $ ./hoshimeguri1 ../data.txt

package main

import (
  "bytes"
  "bufio"
  "fmt"
  "os"
  "strings"
)

func solve(input string) string {
  star := map[byte] map[rune] byte {
    'A': {'W': 'I', 'R': 'H'},
    'I': {'W': 'G', 'R': 'F'},
    'G': {'W': 'E', 'R': 'D'},
    'E': {'W': 'C', 'R': 'B'},
    'C': {'W': 'A', 'R': 'J'},
    'H': {'W': 'C', 'R': 'J'},
    'J': {'W': 'E', 'R': 'B'},
    'B': {'W': 'G', 'R': 'D'},
    'D': {'W': 'I', 'R': 'F'},
    'F': {'W': 'A', 'R': 'H'}}
  point := input[0]
  result := bytes.NewBufferString("")
  result.WriteByte(point)
  for _, c := range input[1:] {
    point = star[point][c]
    result.WriteByte(point)
  }
  return result.String()
}

func test(input string, expected string) {
  actual := solve(input)
  if actual == expected {
    fmt.Print(".")
  } else {
    fmt.Printf(`
input:    %s
expected: %s
actual:   %s
`, input, expected, actual)
  }
}

func main() {
  filename := os.Args[1]
  file, _ := os.Open(filename)
  scanner := bufio.NewScanner(file)
  for scanner.Scan() {
    tokens := strings.Split(scanner.Text(), " ")
    test(tokens[0], tokens[1])
  }
  file.Close()
  fmt.Println()
}
